File: AISuite_QuickStart/app/src/main/java/com/zebra/aisuite_quickstart/java/analyzers/customdetector/mobilenet/TFLiteModel.java
// Copyright 2025 Zebra Technologies Corporation and/or its affiliates. All rights reserved.
package com.zebra.aisuite_quickstart.java.analyzers.customdetector.mobilenet;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Rect;
import android.util.Log;

import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.Tasks;
import com.google.mlkit.vision.common.InputImage;
import com.zebra.ai.vision.entity.DetectionEntity;

import org.tensorflow.lite.Interpreter;
import org.tensorflow.lite.support.common.FileUtil;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * MobileNet SSD v1 COCO object detector (TFLite, quantized uint8).
 * Apache-2.0 licensed model from the TensorFlow Lite object detection examples.
 *
 * Input : [1, 300, 300, 3] uint8 NHWC, RGB values 0–255.
 * Outputs (TFLite_Detection_PostProcess — NMS is baked in):
 *   0 — locations  [1, max_det, 4]  float32  [ymin, xmin, ymax, xmax] normalised 0–1
 *   1 — classes    [1, max_det]     float32  class index (0-based into labelmap)
 *   2 — scores     [1, max_det]     float32  confidence (descending)
 *   3 — num_det    [1]              float32  number of valid detections
 * max_det is read from the model at runtime; boxes are mapped to bitmap pixel space on output.
 */
public class TFLiteModel implements AutoCloseable {

    private static final String TAG        = "TFLiteModel";
    private static final String MODEL_FILE = "models/mobilenet_ssd.tflite";
    private static final float  CONF_THRESHOLD = 0.5f;

    private final Interpreter interpreter;
    private final int maxDetections;
    private final int inputSize;
    private final int boxesIdx;
    private final int scoresIdx;
    private final int numDetIdx;

    // ── Model Loading ────────────────────────────────────────────────────────

    public TFLiteModel(Context context) throws IOException {
        interpreter   = new Interpreter(FileUtil.loadMappedFile(context, MODEL_FILE));
        inputSize     = interpreter.getInputTensor(0).shape()[1];
        int[] idx     = resolveOutputIndices();
        boxesIdx      = idx[0];
        scoresIdx     = idx[1];
        numDetIdx     = idx[2];
        maxDetections = idx[3];
        Log.i(TAG, "Model ready — inputSize=" + inputSize + " maxDetections=" + maxDetections);
    }

    private int[] resolveOutputIndices() {
        int foundBoxes = 0, foundScores = 2, foundNumDet = 3, foundMaxDet = 10;
        int numOutputs = interpreter.getOutputTensorCount();
        for (int i = 0; i < numOutputs; i++) {
            int[]  sh = interpreter.getOutputTensor(i).shape();
            String nm = interpreter.getOutputTensor(i).name();
            Log.i(TAG, "  output[" + i + "] name=" + nm + " shape=" + Arrays.toString(sh)
                    + " dtype=" + interpreter.getOutputTensor(i).dataType());
            if (sh.length == 3 && sh[sh.length - 1] == 4) { foundBoxes = i; foundMaxDet = sh[sh.length - 2]; }
            if (sh.length == 1) foundNumDet = i;
        }
        foundScores = (foundBoxes == 2) ? 0 : 2;
        return new int[]{ foundBoxes, foundScores, foundNumDet, foundMaxDet };
    }

    // ── Inference ────────────────────────────────────────────────────────────

    public Task<List<DetectionEntity>> process(InputImage inputImage) {
        Bitmap bitmap = inputImage.getBitmapInternal();
        List<DetectionEntity> result = detect(bitmap).stream()
                .map(rect -> new DetectionEntity(rect, new ArrayList<>()))
                .collect(java.util.stream.Collectors.toList());
        return Tasks.forResult(result);
    }

    public List<Rect> detect(Bitmap bitmap) {
        ByteBuffer input = toInputBuffer(bitmap);
        int bw = bitmap.getWidth();
        int bh = bitmap.getHeight();

        float[][][] boxes    = new float[1][maxDetections][4];
        float[][]   classes  = new float[1][maxDetections];
        float[][]   scores   = new float[1][maxDetections];
        float[]     numDet   = new float[1];

        Map<Integer, Object> outputs = new HashMap<>();
        outputs.put(boxesIdx,  boxes);
        outputs.put(1,         classes);
        outputs.put(scoresIdx, scores);
        outputs.put(numDetIdx, numDet);

        interpreter.runForMultipleInputsOutputs(new Object[]{input}, outputs);

        int n = Math.min((int) numDet[0], maxDetections);
        List<Rect> results = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            float conf = scores[0][i];
            if (conf < CONF_THRESHOLD) continue;
            float ymin = boxes[0][i][0], xmin = boxes[0][i][1];
            float ymax = boxes[0][i][2], xmax = boxes[0][i][3];
            results.add(new Rect(
                    Math.round(xmin * bw), Math.round(ymin * bh),
                    Math.round(xmax * bw), Math.round(ymax * bh)));
        }
        if (n > 0) {
            Log.v(TAG, "  MobileNet SSD detections kept=" + results.size() + "/" + n
                    + " top_score=" + scores[0][0]);
        }
        return results;
    }

    // ── Pre-processing ───────────────────────────────────────────────────────

    private ByteBuffer toInputBuffer(Bitmap bitmap) {
        Bitmap sw = bitmap.getConfig() == Bitmap.Config.HARDWARE
                ? bitmap.copy(Bitmap.Config.ARGB_8888, false) : bitmap;
        Bitmap scaled = Bitmap.createScaledBitmap(sw, inputSize, inputSize, true);
        int[] pixels = new int[inputSize * inputSize];
        scaled.getPixels(pixels, 0, inputSize, 0, 0, inputSize, inputSize);

        ByteBuffer bufOut = ByteBuffer.allocateDirect(inputSize * inputSize * 3);
        bufOut.order(ByteOrder.nativeOrder());
        for (int px : pixels) {
            bufOut.put((byte) ((px >> 16) & 0xFF));
            bufOut.put((byte) ((px >>  8) & 0xFF));
            bufOut.put((byte) (px & 0xFF));
        }
        bufOut.rewind();
        return bufOut;
    }

    // ── Lifecycle ────────────────────────────────────────────────────────────

    @Override
    public void close() {
        interpreter.close();
    }
}

File: AISuite_QuickStart/app/src/main/java/com/zebra/aisuite_quickstart/kotlin/analyzers/customdetector/mobilenet/TFLiteModel.kt
// Copyright 2025 Zebra Technologies Corporation and/or its affiliates. All rights reserved.
package com.zebra.aisuite_quickstart.kotlin.analyzers.customdetector.mobilenet

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Rect
import android.util.Log
import org.tensorflow.lite.Interpreter
import org.tensorflow.lite.support.common.FileUtil
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * MobileNet SSD v1 COCO object detector (TFLite, quantized uint8).
 * Apache-2.0 licensed model from the TensorFlow Lite object detection examples.
 *
 * Input : [1, 300, 300, 3] uint8 NHWC, RGB values 0–255.
 * Outputs (TFLite_Detection_PostProcess — NMS is baked in):
 *   0 — locations  [1, max_det, 4]  float32  [ymin, xmin, ymax, xmax] normalised 0–1
 *   1 — classes    [1, max_det]     float32  class index (0-based into labelmap)
 *   2 — scores     [1, max_det]     float32  confidence (descending)
 *   3 — num_det    [1]              float32  number of valid detections
 * max_det is read from the model at runtime; boxes are mapped to bitmap pixel space on output.
 */
class TFLiteModel(context: Context) : AutoCloseable {

    companion object {
        private const val TAG            = "TFLiteModel"
        private const val MODEL_FILE     = "models/mobilenet_ssd.tflite"
        private const val CONF_THRESHOLD = 0.5f
    }

    private val interpreter: Interpreter
    private val inputSize: Int
    private val maxDetections: Int
    private val boxesIdx: Int
    private val scoresIdx: Int
    private val numDetIdx: Int

    init {
        Log.d(TAG, "Loading TFLite model from assets: $MODEL_FILE")
        interpreter = Interpreter(FileUtil.loadMappedFile(context, MODEL_FILE))

        // Resolve input size dynamically (typically 300 for MobileNet SSD)
        inputSize = interpreter.getInputTensor(0).shape()[1]

        var foundBoxes  = 0
        var foundScores = 2
        var foundNumDet = 3
        var foundMaxDet = 10

        repeat(interpreter.outputTensorCount) { i ->
            val sh  = interpreter.getOutputTensor(i).shape()
            val nm  = interpreter.getOutputTensor(i).name()
            Log.i(TAG, "  output[$i] name=$nm shape=${sh.contentToString()}" +
                    " dtype=${interpreter.getOutputTensor(i).dataType()}")
            if (sh.size == 3 && sh.last() == 4) { foundBoxes = i; foundMaxDet = sh[sh.size - 2] }
            if (sh.size == 1)                    { foundNumDet = i }
        }
        foundScores = if (foundBoxes == 2) 0 else 2

        boxesIdx      = foundBoxes
        scoresIdx     = foundScores
        numDetIdx     = foundNumDet
        maxDetections = foundMaxDet

        Log.i(TAG, "MobileNet SSD ready — inputSize=$inputSize maxDetections=$maxDetections" +
                " boxesIdx=$boxesIdx scoresIdx=$scoresIdx numDetIdx=$numDetIdx" +
                " inputDtype=${interpreter.getInputTensor(0).dataType()}")
    }

    fun detect(bitmap: Bitmap): List<Rect> {
        val input = toInputBuffer(bitmap)
        val bw = bitmap.width
        val bh = bitmap.height

        val boxes   = Array(1) { Array(maxDetections) { FloatArray(4) } }
        val classes = Array(1) { FloatArray(maxDetections) }
        val scores  = Array(1) { FloatArray(maxDetections) }
        val numDet  = FloatArray(1)

        val outputs: MutableMap<Int, Any> = mutableMapOf(
            boxesIdx  to boxes,
            1         to classes,
            scoresIdx to scores,
            numDetIdx to numDet
        )
        interpreter.runForMultipleInputsOutputs(arrayOf(input), outputs)

        val n = minOf(numDet[0].toInt(), maxDetections)
        val results = (0 until n)
            .filter { scores[0][it] >= CONF_THRESHOLD }
            .map { i ->
                val ymin = boxes[0][i][0]; val xmin = boxes[0][i][1]
                val ymax = boxes[0][i][2]; val xmax = boxes[0][i][3]
                Rect((xmin * bw).toInt(), (ymin * bh).toInt(),
                     (xmax * bw).toInt(), (ymax * bh).toInt())
            }
        if (n > 0) Log.v(TAG, "  MobileNet SSD detections kept=${results.size}/$n top_score=${scores[0][0]}")
        return results
    }

    private fun toInputBuffer(bitmap: Bitmap): ByteBuffer {
        val sw = if (bitmap.config == Bitmap.Config.HARDWARE)
            bitmap.copy(Bitmap.Config.ARGB_8888, false) else bitmap
        val scaled = Bitmap.createScaledBitmap(sw, inputSize, inputSize, true)
        val pixels = IntArray(inputSize * inputSize)
        scaled.getPixels(pixels, 0, inputSize, 0, 0, inputSize, inputSize)

        val buf = ByteBuffer.allocateDirect(inputSize * inputSize * 3)
            .order(ByteOrder.nativeOrder())
        for (px in pixels) {
            buf.put(((px shr 16) and 0xFF).toByte())
            buf.put(((px shr  8) and 0xFF).toByte())
            buf.put((px and 0xFF).toByte())
        }
        buf.rewind()
        return buf
    }

    override fun close() = interpreter.close()
}
